blockindex/sgd_indexer.go

```diff
@@ -7,50 +7,448 @@ package blockindex
+				"inputs": [
+					{
+						"indexed": false,
+						"internalType": "address",
+						"name": "contractAddress",
+						"type": "address"
+					},
+					{
+						"indexed": false,
+						"internalType": "address",
+						"name": "recipient",
+						"type": "address"
+					}
+				],
+				"name": "ContractRegistered",
+				"type": "event"
+			},
+			{
+				"anonymous": false,
+				"inputs": [
+					{
+						"indexed": false,
+						"internalType": "address",
+						"name": "contractAddress",
+						"type": "address"
+					}
+				],
+				"name": "ContractRemoved",
+				"type": "event"
+			},
+			{
+				"anonymous": false,
+				"inputs": [
+					{
+						"indexed": false,
+						"internalType": "address",
+						"name": "contractAddress",
+						"type": "address"
+					}
+				],
+				"name": "ContractDisapproved",
+				"type": "event"
+			},
+			{
+				"anonymous": false,
+				"inputs": [
+					{
+						"indexed": false,
+						"internalType": "uint8",
+						"name": "version",
+						"type": "uint8"
+					}
+				],
+				"name": "Initialized",
+				"type": "event"
+			},
+			{
+				"anonymous": false,
+				"inputs": [
+					{
+						"indexed": true,
+						"internalType": "address",
+						"name": "previousOwner",
+						"type": "address"
+					},
+					{
+						"indexed": true,
+						"internalType": "address",
+						"name": "newOwner",
+						"type": "address"
+					}
+				],
+				"name": "OwnershipTransferred",
+				"type": "event"
+			},
+			{
+				"anonymous": false,
+				"inputs": [
+					{
+						"indexed": true,
+						"internalType": "address",
+						"name": "implementation",
+						"type": "address"
+					}
+				],
+				"name": "Upgraded",
+				"type": "event"
+			}
+]`
+)
+
+func init() {
+	var err error
+	_sgdABI, err = abi.JSON(strings.NewReader(_sgdContractInterfaceABI))
+	if err != nil {
+		panic(err)
+	}
+}
+
+const (
+	_sgdBucket     = "sg"
+	_sgdToHeightNS = "hh"
+	//TODO (millken): currently we fix the percentage to 30%, we can make it configurable in the future
+	_sgdPercentage = uint64(30)
+)
+
+var _sgdCurrentHeight = []byte("currentHeight")
+
 type (
 	// SGDRegistry is the interface for Sharing of Gas-fee with DApps
 	SGDRegistry interface {
 		blockdao.BlockIndexer
 		// CheckContract returns the contract's eligibility for SGD and percentage
 		CheckContract(context.Context, string) (address.Address, uint64, bool, error)
+		// FetchContracts returns all contracts that are eligible for SGD
+		FetchContracts(context.Context) ([]*indexpb.SGDIndex, error)
 	}
 
 	sgdRegistry struct {
+		contract string
+		kvStore  db.KVStore
 	}
 )
 
+func newSgdIndex(contract, receiver []byte) *indexpb.SGDIndex {
+	return &indexpb.SGDIndex{
+		Contract: contract,
+		Receiver: receiver,
+	}
+}
+
 // NewSGDRegistry creates a new SGDIndexer
-func NewSGDRegistry() SGDRegistry {
-	return &sgdRegistry{}
+func NewSGDRegistry(contract string, kv db.KVStore) SGDRegistry {
+	if kv == nil {
+		panic("nil kvstore")
+	}
+	if contract != "" {
```

Comment:
`if contract == ""` should also err out

Reply:
in dev or env, it is allowed to be left blank when we not deployed contracts